pkg/storage/limits/limits.go

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+package limits
+
+import "fmt"
+
+// Any storage is limited, so we have to be sure that data we are going to store would fit underlaying storage structures
+// We could delegate this job to a specific storage plugin, but in this case it has to know too much about job/test descriptor etc
+// Another approach is just fail on insertion of invalid data, but we have to try our best in prevention of this situation
+// (Image a test failing after multiple hours with a error like "you step label is too long!") So as a tradoff we will try
+// to establish and enforce some system wide limits, and make storage plugin developer's life easier
+
+// LimitsValidator provides methods to validate data size from storage perspective
+type LimitsValidator interface {
+	ValidateTestName(testName string) error
+	ValidateTestStepLabel(testStepLabel string) error
+	ValidateEventName(eventName string) error
+	ValidateReporterName(reporterName string) error
+	ValidateJobName(jobName string) error
+	ValidateRequestorName(jobRequestor string) error
+	ValidateServerID(serverID string) error
+
+	// Target name/id??
+}
+
+// ErrParameterIsTooLong implements "error" for generic length check error.
+type ErrParameterIsTooLong struct {
+	DataName  string
+	MaxLen    int
+	ActualLen int
+}
+
+func (err ErrParameterIsTooLong) Error() string {
+	return fmt.Sprintf("%s is too long: %d > %d", err.DataName, err.ActualLen, err.MaxLen)
+}
+
+// Validator is an instance of current storage limits validator
+var Validator LimitsValidator = &BasicStorageLimitsValidator{}
```

Comment:
I would steer away from global variables - can we instead have a `NewValidator` function that returns `(LimitsValidator, error)` instead? Testing is equally easier, and we move away from global state

Reply:
In fact Validator does not have any state.. I just don't like to create instances for each call. We also would need to handle an error every time. I don't see any reason NewValidator could fail, so could it be func NewValidator() LimitsValidator? 
Regarding tests, it would be fairly easy to put mocked validator in to the global var, but I don't see any way to mock value returned by NewValidator, if its called as limits.NewValidator().Validate..(). May be we could use factory here as a tradeoff?

Reply:
a validator as global variable is essentially state, I can use a different term if "state" doesn't feel right. Anyway, subsequent calls depend on the value pointed to by `Validator`, which can be changed, so that such calls may end up using a different object. Testing is an example where this becomes challenging: you cannot run tests in parallel if you want to use different Validator implementations, because you don't have a guarantee that another test hasn't changed it while you're using it

Reply:
I understand problems of shared state, but I don't see a good way to solve them, besides using context propagation, as @xaionaro was pointing out. Factory does not necessarily return a new instance of an object. It returns _some_ instance and caller should not care if it's a new one or not. If I see New... function I expect to get a new instance from it. If I see Get... function I know that I cannot be sure. My understanding of factory is that it's a manageable source of instances, which allows to change instances creation strategy easily. So it would be GetValidator instead of NewValidator. :)

Reply:
I am not outright against singleton factories, but I have one question about it: if the returned instance is global, we lose the ability of running tests in parallel, right? In which case this should be clearly documented at least (i.e. clarify that this is a singleton with all the disadvantages that it carries)

Reply:
Changed to NewValidator.:) I just thought that it would be easier to switch to validation by a storage plugin in future with a single instance of validator, but let's keep it simple for now:)